service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalAttributeReader.java

```diff
@@ -29,20 +55,78 @@ public ExternalAttributeReader(final AttributeCache attributeCache) {
-        return Collections.emptySet();
+    private void setRequestFactory(final OAuth2RestTemplate restTemplate) {
+        CloseableHttpClient httpClient = HttpClientBuilder.create().useSystemProperties().build();
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory(httpClient);
+        requestFactory.setReadTimeout(this.adapterHttpClientTimeout);
+        requestFactory.setConnectTimeout(this.adapterHttpClientTimeout);
+        requestFactory.setConnectionRequestTimeout(this.adapterHttpClientTimeout);
+        restTemplate.setRequestFactory(requestFactory);
+    }
+
+    private OAuth2RestTemplate getAdapterOauth2RestTemplate(
+            final AttributeAdapterConnectionEntity attributeAdapterConnectionEntity) {
+        ClientCredentialsResourceDetails clientCredentials = new ClientCredentialsResourceDetails();
+        clientCredentials.setAccessTokenUri(attributeAdapterConnectionEntity.getUaaTokenUrl());
+        clientCredentials.setClientId(attributeAdapterConnectionEntity.getUaaClientId());
+        clientCredentials.setClientSecret(attributeAdapterConnectionEntity.getUaaClientSecret());
+        OAuth2RestTemplate oAuth2RestTemplate = new OAuth2RestTemplate(clientCredentials);
+        this.setRequestFactory(oAuth2RestTemplate);
+        return oAuth2RestTemplate;
+    }
+
+    private ResponseEntity<AttributesResponse> getAttributesFromAdapter(final String adapterUrl,
+            final HttpEntity adapterEntity, final OAuth2RestTemplate adapterRestTemplate) {
+        return adapterRestTemplate.exchange(adapterUrl, HttpMethod.GET, adapterEntity, AttributesResponse.class);
+    }
+
+    @SuppressWarnings("squid:RedundantThrowsDeclarationCheck")
+    Set<Attribute> getAttributesFromAdapters(final String identifier) throws AttributeRetrievalException {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(PREDIX_ZONE_ID, this.getZone().getName());
+
+        Set<AttributeAdapterConnectionEntity> attributeAdapterConnectionEntities = this
+                .getAttributeAdapterConnections();
+        Set<Attribute> attributes = new HashSet<>();
+
+        for (AttributeAdapterConnectionEntity attributeAdapterConnectionEntity : attributeAdapterConnectionEntities) {
+            OAuth2RestTemplate adapterRestTemplate = this
+                    .getAdapterOauth2RestTemplate(attributeAdapterConnectionEntity);
+            HttpEntity<String> adapterEntity = new HttpEntity<>(headers);
+
+            String adapterUrl = UriComponentsBuilder
+                    .fromUriString(attributeAdapterConnectionEntity.getAdapterEndpoint()).queryParam(ID, identifier)
+                    .toUriString();
+
+            ResponseEntity<AttributesResponse> attributesResponse;
+            try {
+                attributesResponse = this.getAttributesFromAdapter(adapterUrl, adapterEntity, adapterRestTemplate);
+            } catch (final RestClientException e) {
```

Comment:
Done. Also there already seems to be a test that asserts that `Effect.INDETERMINATE` is produced when a `RuntimeException` exception is thrown: `PolicyEvaluationServiceTest#testPolicyEvaluationExceptionHandling`